shared-module/audiomixer/Mixer.c

```diff
@@ -194,32 +152,147 @@ uint32_t add16signed(uint32_t a, uint32_t b) {
 }
 
 uint32_t add16unsigned(uint32_t a, uint32_t b) {
-    #if (defined (__ARM_FEATURE_DSP) && (__ARM_FEATURE_DSP == 1))
-    // Subtract out the DC offset, add and then shift back.
-    a = __USUB16(a, 0x80008000);
-    b = __USUB16(b, 0x80008000);
-    uint32_t sum = __QADD16(a, b);
-    return __UADD16(sum, 0x80008000);
+    #if (defined (__ARM_ARCH_7EM__) && (__ARM_ARCH_7EM__ == 1)) //Cortex-M4 w/FPU
+    return __UHADD16(a, b);
     #else
     uint32_t result = 0;
     for (int8_t i = 0; i < 2; i++) {
         int16_t ai = (a >> (sizeof(uint16_t) * 8 * i)) - 0x8000;
         int16_t bi = (b >> (sizeof(uint16_t) * 8 * i)) - 0x8000;
-        int32_t intermediate = (int32_t) ai + bi;
+        int32_t intermediate = (int32_t) ai + bi / 2;
         if (intermediate > USHRT_MAX) {
             intermediate = USHRT_MAX;
         }
-        result |= ((uint16_t) intermediate + 0x8000) << (sizeof(int16_t) * 8 * i);
+        result |= ((uint32_t) intermediate & 0xffff) << (sizeof(int16_t) * 8 * i);
     }
     return result;
     #endif
 }
 
-audioio_get_buffer_result_t audioio_mixer_get_buffer(audioio_mixer_obj_t* self,
-                                                     bool single_channel,
-                                                     uint8_t channel,
-                                                     uint8_t** buffer,
-                                                     uint32_t* buffer_length) {
+static inline uint32_t mult8unsigned(uint32_t val, int32_t mul) {
+    // if mul == 0, no need in wasting cycles
+    if (mul == 0) {
+        return 0;
+    }
+    /* TODO: workout ARMv7 instructions
+    #if (defined (__ARM_ARCH_7EM__) && (__ARM_ARCH_7EM__ == 1)) //Cortex-M4 w/FPU
+    return val;
+    #else*/
+    mp_printf(&mp_plat_print, "mult8unsigned");
```

Comment:
Debug print snuck in! Will be removed in next push...

Reply:
Please ping when it's gone. :-)

Reply:
Ping! 😄